server/src/main/java/io/crate/analyze/relations/UnionSelect.java

```diff
@@ -109,4 +102,42 @@ public String toString() {
     public boolean isDistinct() {
         return isDistinct;
     }
+
+    private List<ScopedSymbol> unionizeOutputs(List<Symbol> left, List<Symbol> right) {
+        ArrayList<ScopedSymbol> outputs = new ArrayList<>(left.size());
+        for (int i = 0; i < left.size(); i++) {
+            outputs.add(unionizeSymbol(left.get(i), right.get(i)));
+        }
+        return outputs;
+    }
+
+    private ScopedSymbol unionizeSymbol(Symbol left, Symbol right) {
+        return new ScopedSymbol(name, Symbols.pathFromSymbol(left), unionizeTypes(left.valueType(), right.valueType()));
+    }
+
+    private DataType<?> unionizeTypes(DataType<?> leftType, DataType<?> rightType) {
+        DataType<?> type;
+        if (leftType.id() == ObjectType.ID) {
+            type = unionizeObjectType((ObjectType) leftType, (ObjectType) rightType);
+        } else if (leftType.id() == ArrayType.ID) {
+            type = new ArrayType<>(
+                unionizeTypes(
+                    ((ArrayType<?>) leftType).innerType(),
+                    ((ArrayType<?>) rightType).innerType()));
+        } else {
+            type = leftType.isConvertableTo(rightType, false) ? rightType : leftType;
+        }
+        return type;
+    }
+
+    private ObjectType unionizeObjectType(ObjectType left, ObjectType right) {
+        ObjectType.Builder unionizedObjectBuilder = ObjectType.builder();
+        for (var e : left.innerTypes().entrySet()) {
+            unionizedObjectBuilder.setInnerType(e.getKey(), e.getValue());
+        }
+        for (var e : right.innerTypes().entrySet()) {
+            unionizedObjectBuilder.mergeInnerType(e.getKey(), e.getValue(), this::unionizeTypes);
+        }
+        return unionizedObjectBuilder.build();
+    }
```

Comment:
I think this could go into the `ObjectType` as `ObjectType merge(ObjectType other)`.
Maybe it is even worth to have a `DataTypes.merge(left, right)` to replace the `unionizeTypes`

I'd avoid the `unionize` term. It's a new term we don't use so far and has a different meaning from what it's being used for. Stick to `merge` which is established already.

Reply:
Thank you, done